Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -653,86 +653,390 @@ bool TerrainSystem::GetIsHoleFromFloats(float x, float y, Sampler sampler) const
+        queryPositions.emplace_back(AZ::Vector3(position.GetX() + exactRange, position.GetY(), 0.0f)); // right
+        queryPositions.emplace_back(AZ::Vector3(position.GetX(), position.GetY() + exactRange, 0.0f)); // up
     }
 
-    AZStd::vector<float> heights(directionVectors.size());
-    AZStd::vector<bool> exists(directionVectors.size());
-    GetHeightsSynchronous(directionVectors, sampler, heights, exists);
+    AZStd::vector<float> heights(queryPositions.size());
+    AZStd::vector<bool> exists(queryPositions.size());
 
-    for (size_t i = 0, iteratorIndex = 0; i < inPositions.size(); i++, iteratorIndex += indexStepSize)
+    // We want to query the underlying heights with an EXACT sampler as well, so that we get to the input data that might be
+    // at a much higher frequency than the height query resolution.
+    GetHeightsSynchronous(queryPositions, Sampler::EXACT, heights, exists);
+
+    for (size_t inPosIndex = 0, queryPositionIndex = 0; inPosIndex < inPositions.size(); inPosIndex++, queryPositionIndex += queryCount)
     {
-        directionVectors[iteratorIndex].SetZ(heights[iteratorIndex]);
-        directionVectors[iteratorIndex + 1].SetZ(heights[iteratorIndex + 1]);
-        directionVectors[iteratorIndex + 2].SetZ(heights[iteratorIndex + 2]);
-        directionVectors[iteratorIndex + 3].SetZ(heights[iteratorIndex + 3]);
+        terrainExists[inPosIndex] = true;
+
+        for (size_t querySubindex = 0; querySubindex < queryCount; querySubindex++)
+        {
+            // Combine the output heights with our query positions.
+            queryPositions[queryPositionIndex + querySubindex].SetZ(heights[queryPositionIndex + querySubindex]);
 
-        normals[i] = (directionVectors[iteratorIndex + 2] - directionVectors[iteratorIndex + 1]).
-                         Cross(directionVectors[iteratorIndex + 3] - directionVectors[iteratorIndex]).GetNormalized();
+            // We're querying at a much higher frequency than our query resolution, so we'll simply say that the terrain
+            // exists only if all 4 points of the + sign exist.
+            terrainExists[inPosIndex] = terrainExists[inPosIndex] && exists[queryPositionIndex + querySubindex];
+        }
 
-        // This needs better logic for handling cases where some points exist and some don't, but for now we'll say that if
-        // any of the four points exist, then the terrain exists.
-        terrainExists[i] = exists[iteratorIndex] || exists[iteratorIndex + 1] || exists[iteratorIndex + 2] || exists[iteratorIndex + 3];
+        if (terrainExists[inPosIndex])
+        {
+            // We have 4 vertices that make a + sign, cross the two lines to get the normal at the center.
+            // (right - left) x (up - down)
+            normals[inPosIndex] = (queryPositions[queryPositionIndex + 2] - queryPositions[queryPositionIndex + 1])
+                                .Cross(queryPositions[queryPositionIndex + 3] - queryPositions[queryPositionIndex])
```

Comment:
Small suggestion: probably do the same as in BILINEAR that it defines up/down/right/left in CalculateNormal, so we know which index it represents.